source/base/utilities.cc

```diff
@@ -91,6 +90,169 @@ namespace Utilities
 
 
 
+  template <typename Integer, int dim, typename Number>
+  std::vector<std::array<Integer, dim>>
+  inverse_Hilber_space_filling_curve(
+    const std::vector<Point<dim, Number>> &points,
+    const int                              bit_per_dim)
+  {
+    // get bounding box:
+    Point<dim, Number> bl = points[0], tr = points[0];
+    for (const auto &p : points)
+      for (unsigned int d = 0; d < dim; ++d)
+        {
+          const double cid = p[d];
+          if (cid < bl[d])
+            bl[d] = cid;
+          if (cid > tr[d])
+            tr[d] = cid;
+        }
+
+    const Tensor<1, dim, Number> extent = tr - bl;
+
+    // make sure our conversion from fractional coordinates to
+    // Integers work as expected, namely our cast (Number)max_int
+    const int min_bits =
+      std::min(bit_per_dim,
+               std::min(std::numeric_limits<Integer>::digits,
+                        std::numeric_limits<Number>::digits));
+
+    // based on that get the maximum integer:
+    const Integer max_int = (min_bits == std::numeric_limits<Integer>::digits ?
+                               std::numeric_limits<Integer>::max() :
+                               ((Integer)1 << min_bits) - 1);
+
+    std::vector<std::array<Integer, dim>> int_points(points.size());
+
+    for (unsigned int i = 0; i < points.size(); ++i)
+      {
+        const auto rel = points[i] - bl;
+        // convert into integers:
+        for (unsigned int d = 0; d < dim; ++d)
+          {
+            const Number v = rel[d] / extent[d];
+            Assert(v >= 0. && v <= 1., ExcInternalError());
+            int_points[i][d] = (Integer)(v * (Number)max_int);
```

Comment:
if someone knows a better way to map `[0,1]` to `[0, max_int]` without loosing some precision for `double+unsigned long int` combo, let me know.

Reply:
does the cast round or truncate?

I think otherwise the formula should be good.

Reply:
> does the cast round or truncate?

hm, good question... looks like it rounds down, i.e. 
```
std::cout << (unsigned int) (1.8) << std::endl;
```
gives
```
1
```

but that probably does not matter much here.

Reply:
p.s. 

> I think otherwise the formula should be good.

I was more concerned with the fact that I had to limit `max_int` to 52 bits so that `(Number)max_int` won't return `0`.

Reply:
Can we just stop using C-style casts?

Reply:
> Can we just stop using C-style casts?

yes, we can 😄 updated.